### DISCUSSION TITLE
Best way to create animated image switches/slideshows

### DISCUSSION THREAD

**Discussion by Andre601** [mkdocs/mkdocs#3514] [Q&A]
I saw some wikis offer a feature to switch between multiple images (i.e. to depict variations of one specific thing) and would like to implement such a functionality into my MkDocs made page.

The only thing I know about what they do is, that a lua script is usually used for this functionality.

What would be a good aproach on implementing such a feature into my side?

**Comment by Andre601** [comment]
Anyone?

**Reply by pawamoy** [reply]
Some Javascript code to create a carousel? https://stackoverflow.com/a/64218300/3451029

**Reply by Andre601** [reply]
That answer seems a bit limited in that it focuses on images of the same size... Which in my case would not be guaranteed...

I probs would need to look for and collect images with a specific class here...

**Reply by pawamoy** [reply]
A bit heavy but you could use Bootstrap's carousel: https://getbootstrap.com/docs/4.0/components/carousel/

**Reply by Andre601** [reply]
Also, it does not really explain in the slightest how things like the HTML would need to look like to have this working...

**Reply by Andre601** [reply]
I REALLY don't get this javascript...

I modified it so that it would include `img` elements with a `data-md-component="img-carousel"` attribute (Which should make it much easier to get the elements I need), but... is it even executed?

```javascript
var keepTime = 2000;
console.log("Image Carousel started");
function updateCarousel(img) {
    if (img.carouselRunning) {
        let outstyle = img.carousel[img.carouselIndex % img.carousel.length].style;
        outstyle.visibility = 'hidden';
        outstyle.opacity = 0;
        img.carouselIndex = (img.carouselIndex + 1) % img.carousel.length;
        let instyle = img.carousel[img.carouselIndex % img.carousel.length].style;
        instyle.visibility = 'visible';
        instyle.opacity = 1;
        instyle.position = 'absolute';
    }
    setTimeout(updateCarousel, keepTime, img);
}

function setCarouselEvents(img) {
    img.style.visibility = 'hidden';
    img.style.position = 'absolute';
    img.addEventListener(
        'mouseover', function(e) { this.carousel[0].carouselRunning = false; }
    );
    img.addEventListener(
        'mouseout', function(e) { this.carousel[0].carouselRunning = true; }
    );
}

function setCarousel(img) {
    img.carouselRunning = true;
    setCarouselEvents(img);
    img.carouselIndex = 0;
    setTimeout(updateCarousel, 1, img);
}

let imgs = document.querySelectorAll('img[data-md-component="img-carousel"]');
for (let i = 1; i < imgs.length; i++) {
    let pe = imgs[i].previousElementSibling;
    if (!pe) {
        continue;
    }
    if (pe.nodeName != "img") {
        continue;
    }
    if (imgs[i].carousel) {
        continue;
    }
    if (!pe.carousel) {
        pe.carousel = [ pe ];
        setCarousel(pe);
    }
    pe.carousel.push(imgs[i]);
    imgs[i].carousel = pe.carousel;
    setCarouselEvents(imgs[i]);
}
```
It sucks that there is no proper image carousel thing...
